ql/src/java/org/apache/hadoop/hive/ql/io/avro/AvroGenericRecordReader.java

```diff
@@ -88,22 +88,38 @@ public AvroGenericRecordReader(JobConf job, FileSplit split, Reporter reporter)
       gdr.setExpected(latest);
     }
 
-    if (split.getLength() == 0) {
-      this.isEmptyInput = true;
-      this.start = 0;
-      this.reader = null;
+    org.apache.avro.file.FileReader<GenericRecord> nonFinalReader = null;
+    try {
+      if (split.getLength() == 0) {
+        this.isEmptyInput = true;
+        this.start = 0;
+        this.reader = null;
+      } else {
+        this.isEmptyInput = false;
+        nonFinalReader = new DataFileReader<GenericRecord>(new FsInput(split.getPath(), job), gdr);
+        this.reader = nonFinalReader;
+        this.reader.sync(split.getStart());
+        this.start = reader.tell();
```

Comment:
Could we limit the try/catch for this block?

Reply:
No. Not really. Problematic exceptions I saw were coming from `extractWriterTimezoneFromMetadata` below. 
So maybe we could only wrap
```
      this.writerTimezone = extractWriterTimezoneFromMetadata(job, split, gdr);
      this.writerProleptic = extractWriterProlepticFromMetadata(job, split, gdr);
```
section.

Yet I think it is safer to keep it around everything. E.g. I am not sure if `reader.sync` is guaranteed to not throw. 
Also having it all covered guards as from future changes to this code.

Reply:
Ok. Do I understand correctly that the problem is not with the `DataFileReader` creation, but some other initialization in the `AvroGenericRecordReader`? Basically what we want to do is make sure that we free up the resources if the `AvroGenericRecordReader` creation is unsuccessful?

Then I do not understand the importance of the `nonFinalReader`. Could we just set the reader at creation? Later, if it is set but there is an exception then just close it before rethrowing the exception?

Reply:
We could drop `nonFinalReader` if we just wrapped two `extract*` calls with `try..catch` like this:
```java
      if (split.getLength() == 0) {
        this.isEmptyInput = true;
        this.start = 0;
        this.reader = null;
      } else {
        this.isEmptyInput = false;
        this.reader = new DataFileReader<GenericRecord>(new FsInput(split.getPath(), job), gdr);
        this.reader.sync(split.getStart());
        this.start = reader.tell();
      }
      this.stop = split.getStart() + split.getLength();
      this.recordReaderID = new UID();

    try {
      this.writerTimezone = extractWriterTimezoneFromMetadata(job, split, gdr);
      this.writerProleptic = extractWriterProlepticFromMetadata(job, split, gdr);
    } catch (Exception e) {
      if (this.reader != null) {
        try {
          this.reader.close();
        } catch (Exception closeException) {
          if (closeException != e) {
            e.addSuppressed(closeException);
          }
        }
      }
      throw e;
    }
```

But then we can still have a `close`-leak if `this.reader.sync()` or `this.reader.tell()` throws an exception.

And If I want to wrap everything with `try...catch` I need extra variable.
Specifically this (use `this.reader` instead `nonFinalReader`):
```
    try {
      if (split.getLength() == 0) {
        this.isEmptyInput = true;
        this.start = 0;
        this.reader = null;
      } else {
        this.isEmptyInput = false;
        this.reader = new DataFileReader<GenericRecord>(new FsInput(split.getPath(), job), gdr);
        this.reader.sync(split.getStart());
        this.start = reader.tell();
      }
      this.stop = split.getStart() + split.getLength();
      this.recordReaderID = new UID();

      this.writerTimezone = extractWriterTimezoneFromMetadata(job, split, gdr);
      this.writerProleptic = extractWriterProlepticFromMetadata(job, split, gdr);
    } catch (Exception e) {
      if (this.reader != null) {
        try {
          this.reader.close();
        } catch (Exception closeException) {
          if (closeException != e) {
            e.addSuppressed(closeException);
          }
        }
      }
      throw e;
    }
  }
```
does not work because compiler complains that `this.reader` in `catch` clause can be not initialized

And this (initialize `this.reader` to null before `if`):
```
    this.reader = null;
    try {
      if (split.getLength() == 0) {
        this.isEmptyInput = true;
        this.start = 0;
        this.reader = null;
      } else {
        this.isEmptyInput = false;
        this.reader = new DataFileReader<GenericRecord>(new FsInput(split.getPath(), job), gdr);
        this.reader.sync(split.getStart());
        this.start = reader.tell();
      }
      this.stop = split.getStart() + split.getLength();
      this.recordReaderID = new UID();

      this.writerTimezone = extractWriterTimezoneFromMetadata(job, split, gdr);
      this.writerProleptic = extractWriterProlepticFromMetadata(job, split, gdr);
    } catch (Exception e) {
      if (this.reader != null) {
        try {
          this.reader.close();
        } catch (Exception closeException) {
          if (closeException != e) {
            e.addSuppressed(closeException);
          }
        }
      }
      throw e;
    }
  }
```
does not work because `this.reader` is assigned more than once.

Reply:
Maybe this?

```
    if (split.getLength() == 0) {
      this.reader = null;
      this.isEmptyInput = true;
    } else {
      this.reader = new DataFileReader<GenericRecord>(new FsInput(split.getPath(), job), gdr);
      this.isEmptyInput = false;
    }

    try {
      if (this.isEmptyInput) {
        this.start = 0;
      } else {
        this.reader.sync(split.getStart());
        this.start = reader.tell();
      }
      this.stop = split.getStart() + split.getLength();
      this.recordReaderID = new UID();

      this.writerTimezone = extractWriterTimezoneFromMetadata(job, split, gdr);
      this.writerProleptic = extractWriterProlepticFromMetadata(job, split, gdr);
    } catch (Exception e) {
      if (this.reader != null) {
        try {
          this.reader.close();
        } catch (Exception closeException) {
          if (closeException != e) {
            e.addSuppressed(closeException);
          }
        }
      }
      throw e;
    }
```

Reply:
Sure. That would work. If you prefer to split two `if..else` but not have the extra variable, I can change. Should I.

Reply:
Please

Reply:
Updated :)

Reply:
Reverted import changes done by intellij

Reply:
Thanks @losipiuk!